students/johnpharmd/lesson04/mailroom2.py

```diff
@@ -0,0 +1,134 @@
+                if not new_response.isnumeric():
+                    send_ty()
+                else:
+                    new_response = int(new_response)
+                    donors_amts[response][1] += new_response
+                    donors_amts[response][2] += 1
+                    print('Added to', response, '\'s Donations:',
+                    new_response, '\n')
+            elif response not in donors_amts:
+                title = input('Title: "Ms." or "Mr."?: ')
+                new_response = input('Enter a Donation amount' +
+                    ' (in USD): ')
+                if not new_response.isnumeric():
+                    send_ty()
+                else:
+                    new_response = int(new_response)
+                    print('Added to list of Donors:', title,
+                        response, new_response)
+                    donors_amts[response] = ([(title, ), new_response, 1])
+            title = str(donors_amts.get(response)[0])
+            title = title.strip('(').strip(')').strip(',')
+            title = title.strip('\'')
+            donor_dict = {'title': title,
+                    'last_name': response, 'donation': new_response}
+            print('Dear {title} {last_name}, Thank you for your generous donation in the amount of {donation} USD.'.format(**donor_dict))
+            print()
+    program_run()
+
+
+def get_report():
+    print()
+    psv = ['Donor Name', '| Total Given', '| Num Gifts',
+        '| Average Gift']
+    print('{:<15}{:>12}{:>12}{:>12}'.format(psv[0], psv[1],
+        psv[2], psv[3]))
+    for i in range(55):
+        print('-', end='')
+    print()
+    for donor in donors_amts:
+        d1 = donors_amts[donor][1]
+        d2 = donors_amts[donor][2]
+        print('{:<15}{}{:>10}{:>12}{}{:>11}'.format(donor, '  $',
+            d1, d2, '  $', d1 // d2))
+    print()
+    program_run()
+
+
+def send_letters():
+    global donors_amts
+    d_a = donors_amts
+    now = datetime.datetime.now()
+    for donor in donors_amts:
+        title = str(d_a.get(donor)[0])
+        title = title.strip('(').strip(')').strip(',').strip('\'')
+        with open(donor + str(now.year) + str(now.month) +
+            str(now.day) + '.txt', 'w') as of:
+            donor_dict = {'title': title,
+                    'last_name': donor, 'donations': d_a.get(donor)[1]}
+            of.write(
+                'Dear {title} {last_name},\nThank you for your generous previous giving in the amount of {donations} USD.'.format(**donor_dict))
+            of.write('\nAttached is our most recent independent,'
+                + ' third party audit.\n')
+            of.write('\nWe hope you agree that we have been good'
+                + ' stewards of our donors\' funds,\nand that'
+                + ' you will consider donating'
+                + ' again to our project.\n')
+            of.write('\nBest wishes for continued success,')
+            of.write('\n[Signature]')
+            of.close()
```

Comment:
the `with` statement will close the file for you so no need to `of.close()` here